flink-autoscaler/src/main/java/org/apache/flink/autoscaler/ParallelismAdjuster.java

```diff
@@ -0,0 +1,148 @@
+        // we try to adjust the parallelism such that it divides
+        // the numKeyGroupsOrPartitions without a remainder => data is evenly spread across subtasks
+        for (int p = newParallelism; p <= upperBoundForAlignment; p++) {
+            if (numKeyGroupsOrPartitions % p == 0
+                    ||
+                    // When Mode is MAXIMIZE_UTILISATION , Try to find the smallest parallelism
+                    // that can satisfy the current consumption rate.
+                    (mode == MAXIMIZE_UTILISATION
+                            && numKeyGroupsOrPartitions / p
+                                    < numKeyGroupsOrPartitions / newParallelism)) {
+                return p;
+            }
+        }
+
+        // When adjusting the parallelism after rounding up cannot
+        // find the right degree of parallelism to meet requirements,
+        // Try to find the smallest parallelism that can satisfy the current consumption rate.
+        int p =
+                calculateMinimumParallelism(
+                        numKeyGroupsOrPartitions, newParallelism, parallelismLowerLimit);
+        var message =
+                String.format(
+                        SCALE_LIMITED_MESSAGE_FORMAT,
+                        vertex,
+                        newParallelism,
+                        p,
+                        numKeyGroupsOrPartitions,
+                        upperBound,
+                        parallelismLowerLimit);
+        eventHandler.handleEvent(
+                context,
+                AutoScalerEventHandler.Type.Warning,
+                SCALING_LIMITED,
+                message,
+                SCALING_LIMITED + vertex + newParallelism,
+                context.getConfiguration().get(SCALING_EVENT_INTERVAL));
+        return p;
+    }
+
+    private static int calculateMinimumParallelism(
+            int numKeyGroupsOrPartitions, int newParallelism, int parallelismLowerLimit) {
+        int p = newParallelism;
+        for (; p > 0; p--) {
+            if (numKeyGroupsOrPartitions / p > numKeyGroupsOrPartitions / newParallelism) {
+                if (numKeyGroupsOrPartitions % p != 0) {
+                    p++;
+                }
+                break;
+            }
+        }
+        p = Math.max(p, parallelismLowerLimit);
+        return p;
+    }
+
+    /** The mode of the key group or parallelism adjustment. */
+    public enum KeyGroupOrPartitionsAdjustMode implements DescribedEnum {
+        DEFAULT(
+                "This mode ensures that the parallelism adjustment attempts to evenly distribute data across subtasks"
+                        + ". It is particularly effective for source vertices that are aware of partition counts or vertices after "
+                        + "'keyBy' operation. The goal is to have the number of key groups or partitions be divisible by the set parallelism, ensuring even data distribution and reducing data skew."),
```

Comment:
Can `DEFAULT` be renamed to `ABSOLUTELY_EVENLY_DISTRIBUTION`?

It's more clear for users, and we may update the default value for `scaling.key-group.partitions.adjust.mode` option in the future. For example, `MAXIMIZE_UTILISATION` may be more reasonable in the future. And `KeyGroupOrPartitionsAdjustMode.DEFAULT` is a little wired if it's not the default enum.

Reply:
ok，great advice.